### ISSUE TITLE
cotila::sum produces 0.0 on floating point vectors

### ISSUE THREAD

**Issue by 8W9aG** [calebzulawski/cotila#37]
To reproduce, I have this function for computing a vector of gaussian filter weights:

```
constexpr const auto EULER_NUMBER = 2.71828182845904523536;

template <std::size_t N> constexpr cotila::vector<float, N> generateWeights() {
    constexpr const float sigma = (((float(N) / 2.0f) - 1.0f) / 3.0f) + (1.0f / 3.0f);
    constexpr const auto offset = N / 2;
    constexpr const auto c1 = 1.0 / (sigma * cotila::sqrt(2.0 * M_PI));
    constexpr const auto c2 = 2.0 * cotila::exponentiate(sigma, 2.0);
    constexpr const auto weights = cotila::generate<N>([&](int i) -> float {
        return c1 * cotila::exponentiate(EULER_NUMBER, -cotila::exponentiate(float(i) - float(offset), 2.0f) / c2);
    });
    constexpr const float total_weigths = cotila::sum(weights);
    return weights * total_weigths;
}
```

The weights here are as follows:

```
weights	const cotila::vector<float, 31>	
array	float [31]	
[0]	float	0.00141423533
[1]	float	0.00384429004
[2]	float	0.00384429004
[3]	float	0.010449864
[4]	float	0.010449864
[5]	float	0.0284056757
[6]	float	0.0284056757
[7]	float	0.0284056757
[8]	float	0.0772146284
[9]	float	0.0772146284
[10]	float	0.0772146284
[11]	float	0.0772146284
[12]	float	0.0772146284
[13]	float	0.0772146284
[14]	float	0.0772146284
[15]	float	0.0772146284
[16]	float	0.0772146284
[17]	float	0.0772146284
[18]	float	0.0772146284
[19]	float	0.0772146284
[20]	float	0.0772146284
[21]	float	0.0772146284
[22]	float	0.0772146284
[23]	float	0.0284056757
[24]	float	0.0284056757
[25]	float	0.0284056757
[26]	float	0.010449864
[27]	float	0.010449864
[28]	float	0.00384429004
[29]	float	0.00384429004
[30]	float	0.00141423533
```

The `total_weights` are reported as:
`total_weigths	const float	0`

I suspect this is because of the sum accumulator starts at 0 as an integer even with a floating point, and because these float values are low it never reaches past 1. [See here](https://github.com/calebzulawski/cotila/blob/master/include/cotila/vector/math.h#L100)

**Comment by drubinstein** [comment]
I believe I found the issue. 

The return type [here](https://github.com/calebzulawski/cotila/blob/master/include/cotila/vector/utility.h#L43)  should be T not U

**Comment by calebzulawski** [comment]
I think the return type should definitely still be U, but `sum` needs to use a float return type for floats in all cases

**Comment by drubinstein** [comment]
Then would it be better to static cast the 0 in sum to T?

**Comment by calebzulawski** [comment]
Yeah, I think so

**Comment by calebzulawski** [comment]
Closed by #38

**Comment by 8W9aG** [comment]
Thanks a lot!
